pkg/reconciler/certificate/controller.go

```diff
@@ -64,6 +67,12 @@ func NewController(
 	knCertificateInformer.Informer().AddEventHandler(certHandler)
 
 	cmCertificateInformer.Informer().AddEventHandler(controller.HandleAll(impl.EnqueueControllerOf))
+	svcInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
```

Comment:
I am not sure why we need this global sync when any HTTP challenge is changed?

Reply:
We need to reconcile the certificate when the challenge service is ready but you're right that this is too broad. I can change it to a `FilteredGlobalResync` and just enqueue the certificate that matches the service's domain label.

Reply:
I see. I feel it would be a bit difficult and complex to find out the certificate that matches the service's domain label as you have to do some lookups from either cache or storage.

And alternative way is to use [`Tracker`](https://github.com/knative/serving/blob/master/pkg/reconciler/route/route.go#L389) to track the certificate for the HTTP01 service.

The caveat is that in order to make this work, we need to make sure those services exist in the https://github.com/knative/serving/pull/5186/files?file-filters%5B%5D=.go#diff-e2259688f83ba9787c35d121c3cfd39bR210. If they don't exist yet, we need to return an error so that the certificate will be re-enqueued to catch up those services when they are created. Does this make sense?

Reply:
I think I get it. In the case where the service doesn't exist we need to return an error in `setHTTP01Challenges` so that `reconcile` fails which would cause the cert to get re-enqueued until the service does exist. At that point we can track changes and only reconcile the cert tied to that challenge. The only thing is we have to make sure not to call this for certs that are using DNS01 challenges.

Reply:
Yes, exactly.